File: src/Safir.EventSourcing.Abstractions/EventStore.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using JetBrains.Annotations;
using Safir.Messaging;

namespace Safir.EventSourcing
{
    [PublicAPI]
    public abstract class EventStore : IEventStore
    {
        public abstract Task AddAsync<TAggregateId, TId>(
            TAggregateId aggregateId,
            IEvent @event,
            CancellationToken cancellationToken = default);

        public virtual Task AddAsync<TAggregateId, TId>(
            TAggregateId aggregateId,
            IEnumerable<IEvent> events,
            CancellationToken cancellationToken = default)
        {
            return Task.WhenAll(events.Select(x => AddAsync<TAggregateId, TId>(aggregateId, x, cancellationToken)));
        }

        public abstract Task<IEvent> GetAsync<TAggregateId, TId>(TId id, CancellationToken cancellationToken = default);

        public virtual async Task<TEvent> GetAsync<TEvent, TAggregateId, TId>(
            TId id,
            CancellationToken cancellationToken = default)
            where TEvent : IEvent
        {
            return (TEvent)await GetAsync<TAggregateId, TId>(id, cancellationToken);
        }

        public virtual IAsyncEnumerable<IEvent> StreamBackwardsAsync<TAggregateId, TId>(
            TAggregateId aggregateId,
            int? count = null,
            CancellationToken cancellationToken = default)
        {
            // TODO: Swapping start/end is inconsistent in impl. For example the DBContext
            // impl doesn't allow start to be after end
            return StreamAsync<TAggregateId, TId>(aggregateId, int.MaxValue, count ?? 0, cancellationToken);
        }

        public abstract IAsyncEnumerable<IEvent> StreamAsync<TAggregateId, TId>(
            TAggregateId aggregateId,
            int startPosition = 0,
            int endPosition = int.MaxValue,
            CancellationToken cancellationToken = default);
    }
}

File: src/dotnet/src/Safir.EventSourcing.FSharp/EventStore.fs
namespace Safir.EventSourcing.FSharp

open System;
open System.Threading
open System.Threading.Tasks
open Safir.EventSourcing
open Safir.Messaging;

[<AbstractClass>]
type EventStore() =
    abstract AddAsync<'TAggregateId> : 'TAggregateId -> IEvent -> CancellationToken -> Task;
        
    abstract GetAsync<'TAggregateId> : Guid -> CancellationToken -> Task<IEvent>;

    interface IEventStore with
        member this.AddAsync<'TAggregateId>
            (
                aggregateId: 'TAggregateId,
                event: IEvent,
                cancellationToken: CancellationToken
            ) : Task =
            this.AddAsync aggregateId event cancellationToken

        member this.AddAsync<'TAggregateId>
            (
                aggregateId: 'TAggregateId,
                events: System.Collections.Generic.IEnumerable<IEvent>,
                cancellationToken: CancellationToken
            ) : Task =
            // TODO: Learn F# so I can inline this
            let addAsync x =
                this.AddAsync aggregateId x cancellationToken

            Task.WhenAll(Seq.map addAsync events)

        member this.GetAsync<'TAggregateId>
            (
                id: Guid,
                cancellationToken: CancellationToken
            ) : Task<IEvent> =
            this.GetAsync<'TAggregateId> id cancellationToken

        member this.GetAsync<'TEvent, 'TAggregateId when 'TEvent :> IEvent>
            (
                id: Guid,
                cancellationToken: CancellationToken
            ) : Task<'TEvent> =
            // this.GetAsync<'TAggregateId> id cancellationToken
            failwith "todo"
            
        member this.StreamAsync(aggregateId, startPosition, endPosition, cancellationToken) = failwith "todo"

        member this.StreamBackwardsAsync(aggregateId, count, cancellationToken) = failwith "todo"
